### PR TITLE
[GraphOptimizer] Extend Reshape sinking pass for binary eltwise ops

### LINKED ISSUE #5247: [GraphOptimizer] Reshape sinking
While Reshape nodes are noops and do not directly contribute to model execution time, having them in a graph might prevent other optimizations/transformations from happening. So, sinking Reshapes (similar to Transposes) and merging them might potentially lead to better optimized graph.

One particular use case I encountered is coming from `OptimizeTransposeIntoReshape` optimization (https://github.com/pytorch/glow/pull/2630). There was a Transpose in a graph which Glow wasn't able to optimize out in a single `glow::optimize` run, so it sank till a specific point and then became a Reshape. On the next `glow::optimize` iteration, while conditions for further Transpose sinking were satisfied, it didn't happen because it was a Reshape now, not Transpose. Having Reshape sinking logic along with Transpose sinking should help to resolve such cases.

So, does it make sense to have Reshape sinking transformations in GraphOptimizer?

Also, I've noticed there is a transformation in Glow which is opposite to Reshape sinking - Reshape hoisting over Clip (https://github.com/pytorch/glow/pull/4074). @jfix71, could you give any comment on why we would do that?

### PR DESCRIPTION
**Pull request by aturetsk**
Summary:
Extend Reshape sinking pass for binary eltwise ops in GraphOptimizer.

Documentation:
N/A

Fixes #5247 

Test Plan:
Added GraphOptz unit test.

### CODE REVIEW

**Comment by aturetsk** [comment]
@jfix71, continuing effort on adding generic sinking logic started in https://github.com/pytorch/glow/pull/5616.

**Comment by aturetsk** [comment]
Ping

**Comment by aturetsk** [comment]
Ping.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -1143,37 +1143,122 @@ bool SinkReshapes::run(Function *F, const CompilationContext &cctx) {
     auto *node = &N;
 
     // Sink Reshape below eltwise nodes.
-    if (!node->isDataParallel() || node->hasSideEffects()) {
-      continue;
-    }
+    if (node->isDataParallel() && !node->hasSideEffects()) {
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:1146]: Why can't we leave this as an early continue? Looks like there's no logic after the `if`.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 2]
```diff
@@ -1143,37 +1143,122 @@ bool SinkReshapes::run(Function *F, const CompilationContext &cctx) {
...
+        if (RQ) {
+          auto ty = F->getParent()->uniqueTypeWithNewShape(
+              RQ->getResult().getType(), LRN->getInput().dims());
+          rhs = F->createQuantize(RQ->getName(), rhs, ty);
+        }
+        if (RS) {
+          auto ty = F->getParent()->uniqueTypeWithNewShape(
+              RS->getResult().getType(), LRN->getInput().dims());
+          rhs = F->createSplat(RS->getName(), ty, RS->getValue());
+        }
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:1240]: All this code seems unnecessarily complex. We have optimizations that do folding of Reshapes into Constants/Splats, and if we have a `Constant -> Quantize -> Reshape` then it'll get constant folded too. Can we not just add a Reshape here and let other optimizations clean it up later? E.g.

```suggestion
        // Reshape RHS to match LHS.
        NodeValue rhs = F->createSplat(nv.getNode()->getName(), nv, LRN->getInput().dims());
```

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 3]
```diff
@@ -1143,37 +1143,122 @@ bool SinkReshapes::run(Function *F, const CompilationContext &cctx) {
...
-                                   RS->getResult().getType()->dims());
-    newRS->setPredicate(node->getPredicate());
-    out.replaceAllUsesOfWith(newRS->getResult());
+        // RHS must be either a Reshape or a Constant (+ Quantize) or Splat.
+        auto *RQ = dyn_cast<QuantizeNode>(nv);
+        auto *RC = dyn_cast<Constant>(RQ ? RQ->getInput() : nv);
+        auto *RS = dyn_cast<SplatNode>(nv);
+        if (!RRN && !RC && !RS) {
+          continue;
+        }
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:1217]: Combined with the below suggestion of not relying on specific nodes -- can we just check if this is a constant chain of operations and fold if so? I.e. we could make `isConstantOperation` from `ConstantFolding.cpp` usable outside of there and call it here, e.g.

```suggestion
        // RHS must be a constant chain if it's not a reshape, to allow for the reshape to get folded into the chain later on.
        if (!RRN || !isConstantOperation(nv)) {
          continue;
        }
```

(We'd need to do something like assume Interpreter is the constant folding backend, but I think that's generally the case anyway...)

**Review by jfix71 — COMMENTED** [review]
Makes sense, but added some comments about making it general, and less complicated.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 3 above)
Comment (aturetsk) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:1217]: If we use `isConstantOperation`, we'll change the algorithm from O(N) to something like O(N^2). But the optimization should indeed benefit from it. Is that ok?

File: tools/ClassGen/NodeGen.cpp [hunk 4]
```diff
@@ -1432,7 +1431,6 @@ int main(int argc, char **argv) {
       .addInput("Input")
       .addInput("Mapping")
       .addResultFromCtorArg()
-      .dataParallel()
```
[the remark this answers is not in the document]
Comment (aturetsk) [reply tools/ClassGen/NodeGen.cpp:1435]: `IntLookupTable` is not really a data parallel operation. We need the whole `Mapping` input in order to produce a single output element.

**Comment by aturetsk** [comment]
@jfix71, can you comment on the following before I proceed with updating the changes/
> If we use isConstantOperation, we'll change the algorithm from O(N) to something like O(N^2). But the optimization should indeed benefit from it. Is that ok?

**Pull request closed without merging**